controllers/metal3.io/baremetalhost_controller.go

```diff
@@ -906,6 +928,54 @@ func (r *BareMetalHostReconciler) actionInspecting(prov provisioner.Provisioner,
 
 	clearError(info.host)
 	info.host.Status.HardwareDetails = details
+
+	// Create HardwareData with the same name and namesapce as BareMetalHost
+	hardwareData := &metal3v1alpha1.HardwareData{}
+	hardwareDataKey := client.ObjectKey{
+		Name:      info.host.Name,
+		Namespace: info.host.Namespace,
+	}
+	hd := &metal3v1alpha1.HardwareData{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "HardwareData",
+			APIVersion: metal3v1alpha1.GroupVersion.String(),
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      info.host.Name,
+			Namespace: info.host.Namespace,
+			// Register the finalizer immediately
+			Finalizers: []string{
+				hardwareDataFinalizer,
+			},
+			OwnerReferences: []metav1.OwnerReference{
+				*metav1.NewControllerRef(info.host, metal3v1alpha1.GroupVersion.WithKind("BareMetalHost")),
+			},
+		},
+		Spec: metal3v1alpha1.HardwareDataSpec{
+			HardwareDetails: details,
+		},
+	}
+
+	err = r.Client.Get(context.Background(), hardwareDataKey, hardwareData)
+	if err == nil {
```

Comment:
Do we also need to consider the failure to obtain hardwareDate(hardwareData exists, but getting failed for some reason)? This case should be requeued to fetch again.

Reply:
perhaps we can assume that if `k8serrors.IsNotFound(err)` is false, then the resource exits but it got failed some other reason ? So, it would look like 
```
	err = r.Client.Get(context.Background(), hardwareDataKey, hardwareData)
	if err == nil || !k8serrors.IsNotFound(err) {
		// hardwareData found and we reached here due to request for another inspection.
		// Delete it before re-creating.
		if controllerutil.ContainsFinalizer(hardwareData, hardwareDataFinalizer) {
			controllerutil.RemoveFinalizer(hardwareData, hardwareDataFinalizer)
		}
		if err := r.Update(context.Background(), hardwareData); err != nil {
			return actionError{errors.Wrap(err, "failed to remove hardwareData finalizer")}
		}
		if err := r.Client.Delete(context.Background(), hd); err != nil {
			return actionError{errors.Wrap(err, "failed to delete hardwareData")}
		}
	}
	// either hardwareData was deleted above, or not found. We need to re-create it
	if err := r.Client.Create(context.Background(), hd); err != nil {
		return actionError{errors.Wrap(err, "failed to create hardwareData")}
	}
```

Reply:
seems good to me